ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PmtTransactionPool.java

```diff
@@ -0,0 +1,113 @@
+package org.hyperledger.besu.ethereum.privacy;
+
+import org.hyperledger.besu.datatypes.Hash;
+import org.hyperledger.besu.ethereum.chain.BlockAddedEvent;
+import org.hyperledger.besu.ethereum.chain.BlockAddedObserver;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class PmtTransactionPool implements BlockAddedObserver {
+  private static final Logger LOG = LogManager.getLogger();
+  private final Map<Hash, PmtTransactionTracker> pmtPool;
+
+  public PmtTransactionPool() {
+    this.pmtPool = new HashMap<>();
+  }
+
+  @Override
+  public void onBlockAdded(final BlockAddedEvent event) {
+    // remove from the map any transactions that went into this block
+    // TODO take into account re-orgs
+    event
+        .getAddedTransactions()
+        .forEach(
+            tx -> {
+              LOG.debug("removing " + pmtPool.containsKey(tx.getHash()));
+              pmtPool.remove(tx.getHash());
+            });
+  }
+
+  public long getMaxMatchingNonce(final String sender, final String privacyGroupId) {
+
+    return pmtPool.values().stream()
```

Comment:
This performs an *O(n)* scan over all transactions. Are we sure the transaction pool is always small?

It this is an issue, we could always have a second data structure that maps from *(sender, privacy group)* to the *set(transaction hashes)* of those users.